packages/babel-helper-create-class-features-plugin/src/fields.ts

```diff
@@ -770,6 +771,38 @@ export function buildFieldsInitNodes(
       needsClassRef = needsClassRef || replaced;
     }
 
+    // if there are `new.target` in static field
+    // we should replace it with `undefined`
+    if (isStatic && isField) {
+      // fix issue #12737
+      prop.traverse({
+        MetaProperty(path: NodePath<t.MetaProperty>) {
+          const meta = path.get("meta");
+          const property = path.get("property");
+          const { scope } = path;
+
+          if (
+            meta.isIdentifier({ name: "new" }) &&
+            property.isIdentifier({ name: "target" })
+          ) {
+            const func = path.findParent(path => {
+              if (path.isClass()) return true;
+              if (path.isFunction() && !path.isArrowFunctionExpression()) {
+                return true;
+              }
+              return false;
+            });
```

Comment:
Rather than going back up in the AST to check what `new.target` refers to, we can do it in `thisContextVisitor`:

```ts
const thisContextVisitor = traverse.visitors.merge([
  {
    ThisExpression(path, state) {
      state.needsClassRef = true;
      path.replaceWith(t.cloneNode(state.classRef));
    },
    MetaProperty(path: NodePath<t.MetaProperty>, state) {
      if (
        state.shouldReplaceNewTarget &&
        path.get("meta").isIdentifier({ name: "new" }) &&
        path.get("property").isIdentifier({ name: "target" })
      ) {
        path.replaceWith(scope.buildUndefinedNode());
      }
    }
  },
  environmentVisitor,
]);
```

This is faster because:
- The visitor is cached; we don't need to recreate it on every `buildFieldsInitNodes` call
- It only visits what's needed, thanks to `environmentVisitor`
- We can transform both `this` and `new.target` in the same pass, since they need to be transformed exactly in the same scopes


Also, can you add a test for `new.target` inside a static block? (I think it should be undefined):
```js
class A {
  constructor() {
    class B {
      static { new.target; }
    }
  }
}
```